ITR/data/base_providers.py

```diff
@@ -151,46 +183,39 @@ def _convert_benchmark_to_series(self, benchmark: IBenchmark, scope: EScope) ->
-        """
-        results = []
-        for bm in self._EI_benchmarks.__getattribute__(str(scope)).benchmarks:
-            results.append(self._convert_benchmark_to_series(bm, scope))
-        with warnings.catch_warnings():
-            # pd.DataFrame.__init__ (in pandas/core/frame.py) ignores the beautiful dtype information adorning the pd.Series list elements we are providing.  Sad!
-            warnings.simplefilter("ignore")
-            df_bm = pd.DataFrame(results)
-        df_bm.index.names = [self.column_config.REGION, self.column_config.SECTOR, self.column_config.SCOPE]
-        return df_bm
-
-    def _get_intensity_benchmarks(self, company_sector_region_info: pd.DataFrame,
-                                  scope: EScope = EScope.S1S2) -> pd.DataFrame:
+    def _get_intensity_benchmarks(self, company_sector_region_scope: pd.DataFrame) -> pd.DataFrame:
         """
         Overrides subclass method
-        returns a Dataframe with production benchmarks per company_id given a region and sector.
-        :param company_sector_region_info: DataFrame with at least the following columns :
+        returns a Dataframe with intensity benchmarks per company_id given a region and sector.
+        :param company_sector_region_scope: DataFrame with at least the following columns :
         ColumnsConfig.COMPANY_ID, ColumnsConfig.SECTOR and ColumnsConfig.REGION
-        :param scope: a scope
         :return: A DataFrame with company and intensity benchmarks per calendar year per row
         """
-        sectors = company_sector_region_info[self.column_config.SECTOR]
-        regions = company_sector_region_info[self.column_config.REGION].copy()
-        benchmark_projection = self._get_projected_intensities(scope)  # TODO optimize performance
-        mask = regions.isin(benchmark_projection.reset_index()[self.column_config.REGION])
-        regions.loc[~mask] = "Global"
-
-        # benchmark_projection has a scope by construction
-        benchmark_projection = benchmark_projection.loc[list(zip(regions, sectors, [scope] * len(sectors)))]
-        benchmark_projection.index = sectors.index
-        return benchmark_projection
+        benchmark_projections = self._EI_df
+        df = company_sector_region_scope[['sector', 'region', 'scope']]
+
+        df = df.merge(benchmark_projections, left_on=['sector','region','scope'], right_index=True, how='left')
+        mask = df.iloc[:, -1].isna()
+        if mask.any():
+            df = df.reset_index()
+            mask = df.iloc[:, -1].isna()
+            benchmark_global = benchmark_projections.loc[:, 'Global', :]
+            df1 = df[mask].iloc[:, 0:4].merge(benchmark_global, left_on=['sector','scope'], right_index=True, how='left')
+            mask1 = df1.iloc[:, -1].isna()
+            df2 = df1[~mask1].copy()
+            df2.region = 'Global'
+            df.loc[df2.index, :] = df2
+            mask2 = df.iloc[:, -1].isna()
+            df3 = df[~mask2]
```

Comment:
it is not easy for me, to understand, what we do here with df1, df2, df3
could we add comments, explaining, what happens?

Reply:
Done and pushed.